libcodechecker/report.py

```diff
@@ -0,0 +1,181 @@
+
+        if line_content == '' and not os.path.isfile(source_file):
+            LOG.debug("Failed to generate report hash.")
+            LOG.debug('%s does not exists!' % source_file)
+
+        file_name = os.path.basename(source_file)
+        msg = main_section.get('message')
+
+        hash_content = [file_name,
+                        check_name,
+                        msg,
+                        line_content,
+                        str(from_col),
+                        str(until_col)]
+
+        hash_from_ctrl_section = True
+        for i, section in enumerate(ctrl_sections):
+            edges = section['edges']
+
+            try:
+                start_range_begin = edges[0]['start'][0]
+                start_range_end = edges[0]['start'][1]
+
+                end_range_begin = edges[0]['end'][0]
+                end_range_end = edges[0]['end'][1]
+
+                if i > 0:
+                    prev = ctrl_sections[i-1]
+                    col_to_append = compare_ctrl_sections(section, prev)
+                    if col_to_append:
+                        begin_col, end_col = col_to_append
+                        hash_content.append(str(begin_col))
+                        hash_content.append(str(end_col))
+                else:
+                    hash_content.append(str(start_range_begin['col']))
+                    hash_content.append(str(start_range_end['col']))
+
+                hash_content.append(str(end_range_begin['col']))
+                hash_content.append(str(end_range_end['col']))
+            except IndexError as iex:
+                # Edges might be empty.
+                hash_from_ctrl_section = False
+
+        # Hash generation from the control sections failed for some reason
+        # use event section positions for hash generation.
+        if not hash_from_ctrl_section:
+            event_sections = [x for x in path if x.get('kind') == 'event']
+
+            for i, section in enumerate(event_sections):
+                loc = section['location']
+                line_num = loc['line']
+                col_num = loc['col']
+                hash_content.append(str(col_num))
+
+        string_to_hash = '|||'.join(hash_content)
+        return hashlib.md5(string_to_hash.encode()).hexdigest()
+
+    except Exception as ex:
+        LOG.error("Hash generation failed")
+        LOG.error(ex)
+        return ''
+
+
+class Report(object):
+    """
+    Just a minimal separation of the main section
+    from the path section for easier skip/suppression handling
+    and result processing.
+    """
+    def __init__(self, main, bugpath):
+        # Dictionary containing checker name, report hash,
+        # main report position, report message ...
+        self.__main = main
+
+        # Dictionary containing bug path related data
+        # with control, event ... sections.
+        self.__bug_path = bugpath
+
+    @property
+    def main(self):
+        return self.__main
+
+    def print_main(self):
```

Comment:
I used the `print_main` only for debugging and I left it there, but I've removed it now.